From ce4158deb60b49c6195a5f84b2766c5fc1a450b8 Mon Sep 17 00:00:00 2001
From: Liam Mark <lmark@codeaurora.org>
Date: Tue, 6 Mar 2018 13:55:08 -0800
Subject: [PATCH 4/7] libion: save errno value

Ensure that the errno value is saved in ion_ioctl before calling ALOGE
in case ALOGE changes the errno which then changes the return value of
ion_ioctl.

Returning the wrong errno from ion_ioctl can be bad because it can cause
functions like ion_is_legacy to not behave properly.

Also apply the same fix to ion_map.

Change-Id: Ib994b465bfaccdf20963cfdc3f67d1cc5ff3cf8e
---
 libion/ion.c | 12 ++++++++----
 1 file changed, 8 insertions(+), 4 deletions(-)

diff --git a/libion/ion.c b/libion/ion.c
index 583612854..e37396892 100644
--- a/libion/ion.c
+++ b/libion/ion.c
@@ -69,8 +69,10 @@ int ion_close(int fd) {
 static int ion_ioctl(int fd, int req, void* arg) {
     int ret = ioctl(fd, req, arg);
     if (ret < 0) {
-        ALOGE("ioctl %x failed with code %d: %s\n", req, ret, strerror(errno));
-        return -errno;
+        int ret_errno = errno;
+
+        ALOGE("ioctl %x failed with code %d: %s\n", req, ret, strerror(ret_errno));
+        return -ret_errno;
     }
     return ret;
 }
@@ -120,8 +122,10 @@ int ion_map(int fd, ion_user_handle_t handle, size_t length, int prot, int flags
     }
     tmp_ptr = mmap(NULL, length, prot, flags, data.fd, offset);
     if (tmp_ptr == MAP_FAILED) {
-        ALOGE("mmap failed: %s\n", strerror(errno));
-        return -errno;
+        int ret_errno = errno;
+
+        ALOGE("mmap failed: %s\n", strerror(ret_errno));
+        return -ret_errno;
     }
     *map_fd = data.fd;
     *ptr = tmp_ptr;
-- 
2.17.1

